test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Caching/NuGetExe/NuGetExe.cs

```diff
@@ -0,0 +1,148 @@
+            = new ConcurrentDictionary<string, Task<NuGetExe>>();
+
+        private NuGetExe(string pathToExe)
+        {
+            PathToExe = pathToExe;
+        }
+
+        public string PathToExe { get; }
+
+        public bool Debug { get; set; }
+
+        public CommandRunnerResult Execute(CachingTestContext context, string args)
+        {
+            var timeout = 60 * 1000 * 1;
+            if (Debug)
+            {
+                args += " --debug -Verbosity detailed";
+                timeout *= 60;
+            }
+
+            return CommandRunner.Run(
+                PathToExe,
+                context.WorkingPath,
+                args,
+                timeOutInMilliseconds: timeout,
+                waitForExit: true,
+                environmentVariables: new Dictionary<string, string>
+                {
+                    { "NUGET_PACKAGES", context.GlobalPackagesPath },
+                    { "NUGET_HTTP_CACHE_PATH", context.IsolatedHttpCachePath }
+                });
+        }
+
+        public string GetHttpCachePath(CachingTestContext context)
+        {
+            var result = Execute(context, "locals http-cache -list");
+
+            var stdout = result.Item2.Trim();
+
+            // Example:
+            //   stdout = http-cache: C:\Users\jver\AppData\Local\NuGet\v3-cache
+            //   path   = C:\Users\jver\AppData\Local\NuGet\v3-cache
+            var path = stdout.Split(new[] { ':' }, 2)[1].Trim();
+
+            return path;
+        }
+
+        public static async Task<NuGetExe> Get320Async()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.2.0/nuget.exe",
+                "nuget.3.2.0.exe");
+        }
+
+        public static async Task<NuGetExe> Get330Async()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.3.0/nuget.exe",
+                "nuget.3.3.0.exe");
+        }
+
+        public static async Task<NuGetExe> Get340RcAsync()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.4.0-rc/nuget.exe",
+                "nuget.3.4.3-rc.exe");
+        }
+
+        public static async Task<NuGetExe> Get343Async()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.4.3/nuget.exe",
+                "nuget.3.4.3.exe");
+        }
+
+        public static async Task<NuGetExe> Get344Async()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.4.4/NuGet.exe",
+                "nuget.3.4.4.exe");
+        }
+
+        public static async Task<NuGetExe> Get350Beta2Async()
+        {
+            return await DownloadNuGetExeAsync(
+                "https://dist.nuget.org/win-x86-commandline/v3.5.0-beta2/NuGet.exe",
```

Comment:
This method is actually used. By default, the nuget.exe that was just build is tested. But I should add RC1 for convenience.